File: languages/java/src/main/java/bit/sdk/Secrets.java
package bit.sdk;

import static bit.sdk.BitwardenClient.throwingFunctionWrapper;

import bit.sdk.schema.Command;
import bit.sdk.schema.Converter;
import bit.sdk.schema.ResponseForSecretIdentifiersResponse;
import bit.sdk.schema.ResponseForSecretResponse;
import bit.sdk.schema.ResponseForSecretsDeleteResponse;
import bit.sdk.schema.SecretCreateRequest;
import bit.sdk.schema.SecretGetRequest;
import bit.sdk.schema.SecretIdentifiersRequest;
import bit.sdk.schema.SecretPutRequest;
import bit.sdk.schema.SecretsCommand;
import bit.sdk.schema.SecretsDeleteRequest;
import java.util.UUID;

public class Secrets {

    private CommandRunner commandRunner;

    Secrets(CommandRunner commandRunner) {
        this.commandRunner = commandRunner;
    }

    public ResponseForSecretResponse get(UUID id) {
        Command command = new Command();
        SecretsCommand secretsCommand = new SecretsCommand();
        SecretGetRequest secretGetRequest = new SecretGetRequest();
        secretGetRequest.setID(id);
        secretsCommand.setGet(secretGetRequest);
        command.setSecrets(secretsCommand);
        return commandRunner.runCommand(command,
            throwingFunctionWrapper(Converter::ResponseForSecretResponseFromJsonString));
    }

    public ResponseForSecretResponse create(String key, String value, String note, UUID organizationId,
        UUID[] projectIds) {
        Command command = new Command();
        SecretsCommand secretsCommand = new SecretsCommand();
        SecretCreateRequest secretCreateRequest = new SecretCreateRequest();
        secretCreateRequest.setKey(key);
        secretCreateRequest.setValue(value);
        secretCreateRequest.setNote(note);
        secretCreateRequest.setOrganizationID(organizationId);
        secretCreateRequest.setProjectIDS(projectIds);
        secretsCommand.setCreate(secretCreateRequest);
        command.setSecrets(secretsCommand);
        return commandRunner.runCommand(command,
            throwingFunctionWrapper(Converter::ResponseForSecretResponseFromJsonString));
    }

    public ResponseForSecretResponse update(UUID id, String key, String value, String note, UUID organizationId,
        UUID[] projectIds) {
        Command command = new Command();
        SecretsCommand secretsCommand = new SecretsCommand();
        SecretPutRequest secretPutRequest = new SecretPutRequest();
        secretPutRequest.setID(id);
        secretPutRequest.setKey(key);
        secretPutRequest.setValue(value);
        secretPutRequest.setNote(note);
        secretPutRequest.setOrganizationID(organizationId);
        secretPutRequest.setProjectIDS(projectIds);
        secretsCommand.setUpdate(secretPutRequest);
        command.setSecrets(secretsCommand);
        return commandRunner.runCommand(command,
            throwingFunctionWrapper(Converter::ResponseForSecretResponseFromJsonString));
    }

    public ResponseForSecretsDeleteResponse delete(UUID[] ids) {
        Command command = new Command();
        SecretsCommand secretsCommand = new SecretsCommand();
        SecretsDeleteRequest secretsDeleteRequest = new SecretsDeleteRequest();
        secretsDeleteRequest.setIDS(ids);
        secretsCommand.setDelete(secretsDeleteRequest);
        command.setSecrets(secretsCommand);
        return commandRunner.runCommand(command,
            throwingFunctionWrapper(Converter::ResponseForSecretsDeleteResponseFromJsonString));
    }

    public ResponseForSecretIdentifiersResponse list(UUID organizationId) {
        Command command = new Command();
        SecretsCommand secretsCommand = new SecretsCommand();
        SecretIdentifiersRequest secretIdentifiersRequest = new SecretIdentifiersRequest();
        secretIdentifiersRequest.setOrganizationID(organizationId);
        secretsCommand.setList(secretIdentifiersRequest);
        command.setSecrets(secretsCommand);
        return commandRunner.runCommand(command,
            throwingFunctionWrapper(Converter::ResponseForSecretIdentifiersResponseFromJsonString));
    }
}

File: languages/ruby/lib/secrets.rb
# frozen_string_literal: true

require 'json'

class SecretsClient
  def initialize(command_runner)
    @command_runner = command_runner
  end

  def get(id)
    command = create_command(get: SecretGetRequest.new(id: id))
    response = run_command(command)

    secrets_response = ResponseForSecretResponse.from_json!(response).to_dynamic

    raise BitwardenError, secrets_response['errorMessage'] if secrets_response['errorMessage']

    secrets_response['data']
  end

  def get_by_ids(ids)
    command = create_command(get_by_ids: SecretsGetRequest.new(ids: ids))
    response = run_command(command)

    secrets_response = ResponseForSecretIdentifiersResponse.from_json!(response).to_dynamic

    raise BitwardenError, secrets_response['errorMessage'] if secrets_response['errorMessage']

    secrets_response['data']['data']
  end

  def create(key, note, organization_id, project_ids, value)
    command = create_command(
      create: SecretCreateRequest.new(
        key: key, note: note, organization_id: organization_id, project_ids: project_ids, value: value
      )
    )
    response = run_command(command)

    secrets_response = ResponseForSecretResponse.from_json!(response).to_dynamic

    raise BitwardenError, 'Error creating secret' if secrets_response['errorMessage']

    secrets_response['data']
  end

  def list(organization_id)
    command = create_command(list: SecretIdentifiersRequest.new(organization_id: organization_id))
    response = run_command(command)

    secrets_response = ResponseForSecretIdentifiersResponse.from_json!(response).to_dynamic

    raise BitwardenError, 'Error getting list of secrets' if secrets_response['errorMessage']

    secrets_response['data']['data']
  end

  def update(id, key, note, organization_id, project_ids, value)
    command = create_command(
      update: SecretPutRequest.new(
        id: id, key: key, note: note, organization_id: organization_id, project_ids: project_ids, value: value
      )
    )
    response = run_command(command)

    secrets_response = ResponseForSecretResponse.from_json!(response).to_dynamic

    raise BitwardenError, secrets_response['errorMessage'] if secrets_response['errorMessage']

    secrets_response['data']
  end

  def delete_secret(ids)
    command = create_command(delete: SecretsDeleteRequest.new(ids: ids))
    response = run_command(command)

    secrets_response = ResponseForSecretsDeleteResponse.from_json!(response).to_dynamic

    raise BitwardenError, secrets_response['errorMessage'] if secrets_response['errorMessage']

    secrets_response['data']['data']
  end

  private

  def create_command(commands)
    SelectiveCommand.new(secrets: SelectiveSecretsCommand.new(commands))
  end

  def run_command(command)
    response = @command_runner.run(command)
    raise BitwardenError, 'Error getting response' if response.nil?

    response
  end
end
